lib/core/Stage.cpp

```diff
@@ -141,4 +143,28 @@ std::string Stage::dot_string(const std::string& prefix) const {
     return fmt::format("{:s}\"{:s}\" [shape=box, color=darkgreen];\n", prefix, get_unique_name());
 }
 
+// Used to get tid from pthread_t
+struct pthread_fake {
+    char offset[720];
+    pid_t tid;
+    void* others;
+};
```

Comment:
I'm not very comfortable with this as a method of getting the `tid`, it's far too dependent on the `pthread` library implementation on a given system.   As far as I can tell this might not even be portable between Linux distributions. 

I think the correct solution is to use `gettid()`, of course this requires it be called from within the thread, but I think this is possible without editing every `main_thread()` function.  Basically in the `start()` function use a lambda (or create a new function and function pointer) to start the thread on a new function - and then from that function call `gettid()`, register the thread_id with `register_tid` followed by `main_thread()`, followed by an register call.   This gives us a space to work directly within the thread, without making changes to the thread itself.  Note that you'll have to be careful about how the `*this` pointer is handled, and you'll have to avoid using this method for MacOS, since `gettid()` is a Linux only system call.